app/models/wpcc/contributions_calendar.rb

```diff
@@ -20,43 +20,33 @@ def initialize(eligible_salary:,
     end
 
     def schedule
-      periods.map do |period, period_percents|
-        period_percents.symbolize_keys!
-        if period_percents[:employee_percent] &&
-           period_percents[:employer_percent]
-
-          calculate_contribution(period,
-                                 period_percents[:employee_percent],
-                                 period_percents[:employer_percent],
-                                 period_percents[:tax_relief_percent])
-        else
-          calculate_contribution(period,
-                                 employee_percent,
-                                 employer_percent,
-                                 period_percents[:tax_relief_percent])
-        end
+      periods_above_user_contributions.map do |period, percents|
+        employee_percentage = percents[:employee_percent] || employee_percent
+        employer_percentage = percents[:employer_percent] || employer_percent
+
+        Wpcc::PeriodContributionCalculator.new(
+          name: period.to_s,
+          employee_percent: employee_percentage,
+          employer_percent: employer_percentage,
+          eligible_salary: eligible_salary,
+          salary_frequency: salary_frequency,
+          tax_relief_percent: percents[:tax_relief_percent]
+        ).contribution
       end
     end
 
     private
 
-    def periods
-      PERIODS
+    def periods_above_user_contributions
+      periods.select do |_, percents|
+        percents[:employee_percent].blank? ||
+          percents[:employee_percent] > employee_percent ||
+          percents[:employer_percent] > employer_percent
+      end
```

Comment:
This is a bit complicated, can it be simplified? or at least defined in a separate private method.